app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt

```diff
@@ -0,0 +1,215 @@
+import androidx.compose.material.IconButton
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.clip
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import com.google.accompanist.insets.navigationBarsPadding
+import com.google.accompanist.insets.statusBarsPadding
+import mozilla.telemetry.glean.private.NoExtras
+import org.mozilla.fenix.GleanMetrics.Onboarding
+import org.mozilla.fenix.R
+import org.mozilla.fenix.theme.FirefoxTheme
+
+/**
+ * Enum that represents the onboarding page that is displayed.
+ */
+private enum class UpgradeOnboardingState {
+    Welcome,
+    SyncSignIn,
+}
+
+/**
+ * A screen for displaying a welcome and sync sign in onboarding.
+ *
+ * @param isUserSignedIn Whether or not the user is signed into their Firefox Sync account.
+ * @param onDismiss Invoked when the user clicks on the close or "Skip" button.
+ * @param onSignInButtonClick Invoked when the user clicks on the "Sign In" button
+ */
+@Composable
+fun UpgradeOnboardingScreen(
+    isUserSignedIn: Boolean,
+    onDismiss: () -> Unit,
+    onSignInButtonClick: () -> Unit,
+) {
+    var onboardingState by remember { mutableStateOf(UpgradeOnboardingState.Welcome) }
+
+    OnboardingContent(
+        onboardingState = onboardingState,
+        onCloseClick = {
+            when (onboardingState) {
+                UpgradeOnboardingState.Welcome -> Onboarding.welcomeCloseClicked.record(NoExtras())
+                UpgradeOnboardingState.SyncSignIn -> Onboarding.syncCloseClicked.record(NoExtras())
+            }
+            onDismiss()
+        },
+        onPrimaryButtonClick = {
+            when (onboardingState) {
+                UpgradeOnboardingState.Welcome -> {
+                    Onboarding.welcomeGetStartedClicked.record(NoExtras())
+                    if (isUserSignedIn) {
+                        onDismiss()
+                    } else {
+                        onboardingState = UpgradeOnboardingState.SyncSignIn
+                    }
+                }
+                UpgradeOnboardingState.SyncSignIn -> {
+                    Onboarding.syncSignInClicked.record(NoExtras())
+                    onSignInButtonClick()
+                }
+            }
+        },
+        onSecondaryButtonClick = {
+            when (onboardingState) {
+                UpgradeOnboardingState.Welcome -> {
+                    // nothing as welcome doesn't have a secondary button
+                }
+                UpgradeOnboardingState.SyncSignIn -> {
+                    Onboarding.syncSkipClicked.record(NoExtras())
```

Comment:
do the metrics pings compile with the Compose previews?

Reply:
This is fair point.

Reply:
I am interested in continuing the discussion of simplifying the Controller/Interactor pattern, so I added it as a topic to our next experience team sync.

Reply:
@MatthewTighe Sounds good to me. I think my ask would always be to start up a RFC - it doesn't have to be long or anything, but we just want to do some due diligence and provide a forum to address any concerns from the wider team.

Reply:
Yeah the previews work, I think maybe because the metrics ping is in the side effects